lte/cloud/go/services/subscriberdb/digest.go

```diff
@@ -52,9 +53,92 @@ func GetDigest(networkID string) (string, error) {
 	return digest, nil
 }
 
+// GetPerSubDigests generates a set of individual subscriber digests ordered by their IDs.
+func GetPerSubDigests(network string) ([]*lte_protos.SubscriberDigestByID, error) {
+	// HACK: apn resources digest is concatenated to every individual subscriber digest
+	// so that the entire set of per-sub digests would capture changes in apn resources
+	apnDigest, err := getApnResourcesDigest(network)
+	if err != nil {
+		glog.Errorf("Failed to generate digest for apn resources of network %+v: %+v", network, err)
+		apnDigest = ""
+	}
+
+	apnsByName, err := LoadApnsByName(network)
+	if err != nil {
+		return nil, err
+	}
+	perSubDigests := []*lte_protos.SubscriberDigestByID{}
+	token := ""
+	foundEmptyToken := false
+	for !foundEmptyToken {
+		// the resultant subProtos lists are already ordered by subscriber id due to the nature
+		// of LoadSubProtosPage
+		subProtos, nextToken, err := LoadSubProtosPage(0, token, network, apnsByName, lte_models.ApnResources{})
+		if err != nil {
+			return nil, err
+		}
+		for _, subProto := range subProtos {
+			digest, err := mproto.HashDeterministic(subProto)
+			if err != nil {
+				// Swallow errors related to a single subscriber digest to avoid affecting the rest
+				glog.Errorf("Failed to generate digest for subscriber %+v of network %+v: %+v", subProto.Sid.Id, network, err)
+				digest = ""
```

Comment:
Hm in this case, the sub's digest is still getting set. So I think, if we fail in any particular sub here, just fail for the entire network

Then, I think it makes sense to add some logic when we're retrieving digests in subscriberdb -- if the "last updated" for either type of digest is outdated by like 3x or more the regen interval, return "out of sync" for safety. Thoughts?

Reply:
Currently the per-sub digest store doesn't store `last_updated` information; this info is stored exclusively in the flat digest store as the only source of truth when it comes to last-updated time. Therefore, if per-sub digest generation failed for a network, should the overall digest generation (including flat digest) just fail for this network, so that the `last_updated` time for this network is not updated to `now()`? @hcgatewood

Reply:
Yeah I think let's go with that -- on a per-network basis, only update the "last_updated" info if you fully+successfully update all digests for that network

Then on the read side of things, we can take into account the "last_updated" and, like I mentioned, if it's too long ago we can say "must resync"

Thoughts?

Reply:
I think that makes sense! In this case, if per-sub digest generation fails, the `last_updated_at` for this network would not update (and will stay at an "outdated" status), so in the next loop a resync will be enforced. @hcgatewood

Reply:
Yes. We just need to be careful in the "SyncSubscribers" rpc to ensure we check the "last updated time" -- if it's too old, SyncSubscribers should *also* return "resync"